sensor/common/compliance/nodescan_handler_test.go

```diff
@@ -89,47 +107,148 @@ func (s *NodeScanHandlerTestSuite) startForwardingToCentral(c <-chan *central.Ms
+	s.NoError(producer.stoppedC.Wait())
+	s.NoError(consumer.stoppedC.Wait())
+
+	errTest := errors.New("example-stop-error")
+	h.Stop(errTest)
+	s.ErrorIs(h.Stopped().Wait(), errTest)
+}
+
+// generateTestInputNoClose generates numToProduce messages of type NodeScanV2.
+// It returns a channel that must be closed by the caller.
+func (s *NodeScanHandlerTestSuite) generateTestInputNoClose(numToProduce int) (chan *storage.NodeScanV2, stoppable) {
+	input := make(chan *storage.NodeScanV2)
+	st := newStoppable()
 	go func() {
-		defer close(nodeScans)
-		for i := 0; i < 4; i++ {
+		defer st.signalStopped()
+		for i := 0; i < numToProduce; i++ {
 			select {
-			case <-s.ctx.Done():
+			case <-st.stopC.Done():
 				return
-			default:
-				nodeScans <- fakeNodeScanV2("Node")
-				if i == 0 {
-					h.Stop(nil)
+			case input <- fakeNodeScanV2(fmt.Sprintf("Node-%d", i)):
+			}
+		}
+	}()
+	return input, st
+}
+
+// consumeAndCount consumes maximally numToConsume messages from the channel and counts the consumed messages
+// It sets error of stoppable.stopC if the number of messages consumed were less than numToConsume
+func consumeAndCount[T any](ch <-chan *T, numToConsume int) stoppable {
+	// simulate Central: consume all messages from h.ResponsesC()
+	st := newStoppable()
+	go func() {
+		defer st.signalStopped()
+		for i := 0; i < numToConsume; i++ {
+			select {
+			case <-st.stopC.Done():
+				st.stopC.Reset()
+				st.stopC.SignalWithError(fmt.Errorf("consumer consumed %d messages but expected to do %d", i, numToConsume))
+				return
+			case _, ok := <-ch:
+				if !ok {
+					st.stopC.SignalWithError(fmt.Errorf("consumer consumed %d messages but expected to do %d", i, numToConsume))
+					return
 				}
 			}
 		}
 	}()
-	err := h.Start()
-	assert.NoError(s.T(), err)
+	return st
 }
 
 func (s *NodeScanHandlerTestSuite) TestRestartHandler() {
-	nodeScans := make(chan *storage.NodeScanV2)
-	defer close(nodeScans)
-	h := NewNodeScanHandler(nodeScans)
+	s.NotPanics(func() {
+		ch, producer := s.generateTestInputNoClose(10)
+		defer close(ch)
+		h := NewNodeScanHandler(ch)
+		s.NoError(h.Start())
+		consumer := consumeAndCount(h.ResponsesC(), 10)
+		s.NoError(producer.stoppedC.Wait())
+		s.NoError(consumer.stoppedC.Wait())
+		h.Stop(nil)
+		s.NoError(h.Stopped().Wait())
 
-	assert.NoError(s.T(), h.Start())
-	h.Stop(nil)
+		s.ErrorIs(h.Start(), errStartMoreThanOnce)
+		s.NoError(h.Stopped().Wait())
+	})
+}
 
-	// try to start & stop the stopped handler again
-	assert.Error(s.T(), h.Start())
-	h.Stop(nil)
+func (s *NodeScanHandlerTestSuite) TestDoubleStartHandler() {
+	s.NotPanics(func() {
+		ch, producer := s.generateTestInputNoClose(10)
+		defer close(ch)
+		h := NewNodeScanHandler(ch)
+		s.NoError(h.Start())
+
+		consumer := consumeAndCount(h.ResponsesC(), 10)
+		s.NoError(producer.stoppedC.Wait())
+
+		s.ErrorIs(h.Start(), errStartMoreThanOnce)
+		h.Stop(nil)
+		s.NoError(h.Stopped().Wait())
+
+		s.NoError(consumer.stoppedC.Wait())
+	})
+}
```

Comment:
This implementation fails test once in a while. Here's what I can suggest
```suggestion
func (s *NodeScanHandlerTestSuite) TestDoubleStartHandler() {
	s.NotPanics(func() {
		ch, producer := s.generateTestInputNoClose(10)
		defer close(ch)
		h := NewNodeScanHandler(ch)

		s.NoError(h.Start())
		s.ErrorIs(h.Start(), errStartMoreThanOnce)

		consumer := consumeAndCount(h.ResponsesC(), 10)

		s.ErrorIs(h.Start(), errStartMoreThanOnce)

		s.NoError(producer.stoppedC.Wait())
		s.NoError(consumer.stoppedC.Wait())

		h.Stop(nil)
		s.NoError(h.Stopped().Wait())
	})
}
```